src/clj/coffi/mem.clj

```diff
@@ -1350,3 +1766,523 @@
+                      ::float  'floats
+                      ::double 'doubles}
+        single-types {::byte     'byte
+                      ::short    'short
+                      ::int      'int
+                      ::long     'long
+                      ::char     'char
+                      ::float    'float
+                      ::double   'double
+                      ::c-string 'String}]
+    (cond (and arr? raw?) (get array-types _type 'objects)
+          (and arr?)      `clojure.lang.IPersistentVector
+          :default        (get single-types _type (keyword (str *ns*) (str _type))))))
+
+(defn- coffitype->array-fn [_type]
+  (get
+   {:coffi.mem/byte   `byte-array
+    :coffi.mem/short  `short-array
+    :coffi.mem/int    `int-array
+    :coffi.mem/long   `long-array
+    :coffi.mem/char   `char-array
+    :coffi.mem/float  `float-array
+    :coffi.mem/double `double-array}
+   _type
+   `object-array))
+
+(defn- coffitype->array-write-fn [_type]
+  ({:coffi.mem/byte   `write-bytes
+    :coffi.mem/short  `write-shorts
+    :coffi.mem/int    `write-ints
+    :coffi.mem/long   `write-longs
+    :coffi.mem/char   `write-chars
+    :coffi.mem/float  `write-floats
+    :coffi.mem/double `write-doubles} _type))
+
+(defn- coffitype->array-read-fn [_type]
+  ({:coffi.mem/byte   `read-bytes
+    :coffi.mem/short  `read-shorts
+    :coffi.mem/int    `read-ints
+    :coffi.mem/long   `read-longs
+    :coffi.mem/char   `read-chars
+    :coffi.mem/float  `read-floats
+    :coffi.mem/double `read-doubles} _type))
+
+(defmulti  generate-deserialize (fn [& xs] (if (vector? (first xs)) (first (first xs)) (first xs))))
+
+(defmethod generate-deserialize :coffi.mem/byte     [_type offset segment-source-form] `(read-byte    ~segment-source-form ~offset))
+(defmethod generate-deserialize :coffi.mem/short    [_type offset segment-source-form] `(read-short   ~segment-source-form ~offset))
+(defmethod generate-deserialize :coffi.mem/int      [_type offset segment-source-form] `(read-int     ~segment-source-form ~offset))
+(defmethod generate-deserialize :coffi.mem/long     [_type offset segment-source-form] `(read-long    ~segment-source-form ~offset))
+(defmethod generate-deserialize :coffi.mem/char     [_type offset segment-source-form] `(read-char    ~segment-source-form ~offset))
+(defmethod generate-deserialize :coffi.mem/float    [_type offset segment-source-form] `(read-float   ~segment-source-form ~offset))
+(defmethod generate-deserialize :coffi.mem/double   [_type offset segment-source-form] `(read-double  ~segment-source-form ~offset))
+(defmethod generate-deserialize :coffi.mem/pointer  [_type offset segment-source-form] `(read-address ~segment-source-form ~offset))
+(defmethod generate-deserialize :coffi.mem/c-string [_type offset segment-source-form] (list `.getString (list `.reinterpret (list `.get (with-meta segment-source-form {:tag 'java.lang.foreign.MemorySegment}) `pointer-layout offset) `Integer/MAX_VALUE) 0))
```

Comment:
Was there a specific thought that led to this version as opposed to the quasiquoted version? Nothing wrong with it, just curious about the thought process.
```suggestion
(defmethod generate-deserialize :coffi.mem/c-string [_type offset segment-source-form]
  `(.getString (.reinterpret (.get ~(with-meta segment-source-form {:tag 'java.lang.foreign.MemorySegment}) pointer-layout ~offset) Integer/MAX_VALUE) 0))
```

Reply:
not really. i *think* this happened because at some point some metadata got swallowed up, and i suspected it ended up on a form or symbol in a macro and not on the emitted form and i wrote it explicitly while eliminating sources of error and never rewrote this. i agree on the syntax quoted version if only for uniformity.